File: LeetCode75_Artur/Two Pointers/11_Container_With_Most_Water.cpp
/*
 https://leetcode.com/problems/container-with-most-water/description/?envType=study-plan-v2&envId=leetcode-75
*/

#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    int maxArea(vector<int>& height)
    {
        int left = 0;
        int right = (int) height.size() - 1;

        int width = (int) height.size() - 1;
        int currVolume;
        int currMax = 0;

        while( left < right )
        {
            currVolume = width * min(height[left], height[right]);
            currMax = max(currVolume, currMax);

            if( height[left] > height[right] )
                right--;
            else
                left++;
            width--;
        }

        return currMax;
    }
};


int main()
{
    Solution solution;

    vector<int> test1 = {1,8,6,2,5,4,8,3,7};
    vector<int> test2 = {1,1};

    cout << solution.maxArea(test1) << endl;
    cout << solution.maxArea(test2) << endl;
}

/*
 Main idea:
 Є два поінтера, один на початку масиву, один в кінці,
 вони відповідають за межі контейнера. Кожного разу посуваємо
 до центру той поінтер, значення на якому є менше, щоб
 максимізувати об'єм.

 Складність роботи алгоритму: O(n)
*/

File: LeetCode75_Artur/Two Pointers/11_Container_With_Most_Water.py
"""
 https://leetcode.com/problems/container-with-most-water/description/?envType=study-plan-v2&envId=leetcode-75
"""
from typing import Optional, List

class Solution:
    def maxArea(self, height: List[int]) -> int:
        left = 0
        right = len(height) - 1
        max_area = 0
        width = len(height) - 1
        
        while left < right:
            curr_area = width * min(height[left], height[right])
            max_area = max(max_area, curr_area)
            
            if height[left] > height[right]:
                right -= 1
            else:
                left += 1
                
            width -= 1
            
        return max_area

    
if __name__ == "__main__":
    test1 = [1,8,6,2,5,4,8,3,7]
    test2 = [1,1]
    
    print(Solution().maxArea(test1))
    print(Solution().maxArea(test2))
